server/src/main/java/org/opensearch/index/translog/RemoteFsTranslog.java

```diff
@@ -423,11 +433,39 @@ protected void setMinSeqNoToKeep(long seqNo) {
         this.minSeqNoToKeep = seqNo;
     }
 
+    @Override
+    protected Releasable drainSync() {
+        try {
+            if (syncPermit.tryAcquire(SYNC_PERMIT, 1, TimeUnit.MINUTES)) {
+                boolean result = pauseSync.compareAndSet(false, true);
+                assert result && syncPermit.availablePermits() == 0;
+                logger.info("All inflight remote translog syncs finished and further syncs paused");
+                return Releasables.releaseOnce(() -> {
+                    syncPermit.release(SYNC_PERMIT);
+                    boolean wasSyncPaused = pauseSync.getAndSet(false);
+                    assert syncPermit.availablePermits() == SYNC_PERMIT : "Available permits is " + syncPermit.availablePermits();
+                    assert wasSyncPaused : "RemoteFsTranslog sync was not paused before re-enabling it";
+                    logger.info("Resumed remote translog sync back on relocation failure");
+                });
+            } else {
+                throw new TimeoutException("Timeout while acquiring all permits");
+            }
+        } catch (TimeoutException | InterruptedException e) {
+            throw new RuntimeException("Failed to acquire all permits", e);
+        }
+    }
+
     @Override
     public void trimUnreferencedReaders() throws IOException {
         // clean up local translog files and updates readers
         super.trimUnreferencedReaders();
 
+        // This is to ensure that after the permits are acquired during primary relocation, there are no further modification on remote
+        // store.
+        if (pauseSync.get()) {
```

Comment:
Thinking if we need a check if primary mode started here as well

Reply:
Made the change.